packages/cubejs-query-orchestrator/src/driver/BaseDriver.ts

```diff
@@ -261,79 +237,72 @@ export class BaseDriver {
+  public dropTable(tableName: string, options?: unknown): Promise<unknown> {
     return this.query(`DROP TABLE ${tableName}`, [], options);
   }
 
-  /**
-   * @param {number} paramIndex
-   * @return {string}
-   */
-  param(_paramIndex) {
+  public param(_paramIndex: number): string {
     return '?';
   }
 
-  testConnectionTimeout() {
+  public testConnectionTimeout() {
     return 10000;
   }
 
-  async downloadTable(table, _options) {
+  public async downloadTable(table: string, _options: ExternalDriverCompatibilities): Promise<DownloadTableMemoryData | DownloadTableCSVData> {
     return { rows: await this.query(`SELECT * FROM ${table}`) };
   }
 
-  async uploadTable(table, columns, tableData) {
-    return this.uploadTableWithIndexes(table, columns, tableData, [], null);
+  public async uploadTable(table: string, columns: TableStructure, tableData: DownloadTableData) {
+    return this.uploadTableWithIndexes(table, columns, tableData, [], null, [], [], []);
   }
 
-  async uploadTableWithIndexes(table, columns, tableData, indexesSql, _uniqueKeyColumns, _queryTracingObj, _aggregates, _createTableIndexes) {
-    if (!tableData.rows) {
+  public async uploadTableWithIndexes(table: string, columns: TableStructure, tableData: DownloadTableData, indexesSql: IndexesSQL, _uniqueKeyColumns: string[] | null, _queryTracingObj: any, _aggregates: string[], _createTableIndexes: CreateTableIndex[]) {
+    if (!isDownloadTableMemoryData(tableData)) {
       throw new Error(`${this.constructor} driver supports only rows upload`);
     }
 
     await this.createTable(table, columns);
     try {
-      for (let i = 0; i < tableData.rows.length; i++) {
-        await this.query(
-          `INSERT INTO ${table}
-        (${columns.map(c => this.quoteIdentifier(c.name)).join(', ')})
-        VALUES (${columns.map((c, paramIndex) => this.param(paramIndex)).join(', ')})`,
-          columns.map(c => this.toColumnValue(tableData.rows[i][c.name], c.type))
-        );
-      }
-      for (let i = 0; i < indexesSql.length; i++) {
-        const [query, params] = indexesSql[i].sql;
-        await this.query(query, params);
+      if (isDownloadTableMemoryData(tableData)) {
+        for (let i = 0; i < tableData.rows.length; i++) {
+          await this.query(
+            `INSERT INTO ${table}
+          (${columns.map(c => this.quoteIdentifier(c.name)).join(', ')})
+          VALUES (${columns.map((c, paramIndex) => this.param(paramIndex)).join(', ')})`,
+            columns.map(c => this.toColumnValue(tableData.rows[i][c.name] as string, c.type))
+          );
+        }
+        for (let i = 0; i < indexesSql.length; i++) {
+          const [query, params] = indexesSql[i].sql;
+          await this.query(query, params);
+        }
       }
     } catch (e) {
       await this.dropTable(table);
       throw e;
     }
   }
 
-  toColumnValue(value, _genericType) {
+  protected toColumnValue(value: string, _genericType: string): string | boolean {
```

Comment:
`GenericDataBaseType`, it's an opaque type, but we can make it as union of constants in the future.